{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Confirmation of known CNV associated with psychosis\n",
    "\n",
    "The largest genome-wide study on the role of CNVs in psychosis, conducted by Marshall et al. in 2016 with 41,321 subjects, identified significant associations at eight loci: ```1q21.1```, ```2p16.3 (NRXN1)```, ```3q29```, ```7q11.2```, ```15q13.3```, ```distal 16p11.2```, ```proximal 16p11.2```, and ```22q11.2```.\n",
    "\n",
    "Identifying the genetic causes of psychosis is crucial for advancing healthcare, as it enables the development of personalized treatments, aids in family planning, and helps reduce self-blame. Additionally, it plays a significant role in diminishing the stigma associated with psychiatric disorders. Consequently, following data preprocessing, the next step involved detecting known CNVs associated with psychosis in the study cohort.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libraries\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the CNV and Marshal data\n",
    "cnv_df = pd.read_csv(\"../results/merged_all_cnv_annotations.csv\")\n",
    "marshal_df = pd.read_csv(\"../databases/marshal_cnv.csv\")\n",
    "\n",
    "# Filter out rows with non protein coding CNVs\n",
    "cnv_df = cnv_df[cnv_df[\"All protein coding genes\"].notna()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This detection was achieved by cross-referencing the CNVs listed by Marshall et al., including their coordinates, cytobands, and CNV types, with the CNVs present in the cohort. Overlaps were filtered to ensure they exceeded a threshold of zero."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After identifying CNVs based on the criteria outlined by Marshall et al., we observed 9 patients with CNVs at 3 distinct loci: ```22q11.21``` (2 patients with SCZ and 2 with BD), ```2p16.3``` (2 patients with SCZ and 1 UHR-NC), and ```1q21.1``` (2 SCZ patients). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>Cytoband</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Type</th>\n",
       "      <th>Overlap with Marshal (bp)</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Brain_genes</th>\n",
       "      <th>is_rare</th>\n",
       "      <th>Genes (Marshal)</th>\n",
       "      <th>Status (Marshal)</th>\n",
       "      <th>Classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S36774</td>\n",
       "      <td>SCZ</td>\n",
       "      <td>22q11.21</td>\n",
       "      <td>chr22</td>\n",
       "      <td>18731658</td>\n",
       "      <td>21110696</td>\n",
       "      <td>DEL</td>\n",
       "      <td>1018342</td>\n",
       "      <td>LOC102724770, DGCR6, LOC102724788, PRODH, DGCR...</td>\n",
       "      <td>PRODH, ESS2, SEPTIN5, GP1BB, ARVCF, RTN4R, PI4...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>risk</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S36820</td>\n",
       "      <td>UHR-NC</td>\n",
       "      <td>2p16.3</td>\n",
       "      <td>chr2</td>\n",
       "      <td>50672899</td>\n",
       "      <td>50921973</td>\n",
       "      <td>DEL</td>\n",
       "      <td>249074</td>\n",
       "      <td>NRXN1</td>\n",
       "      <td>NRXN1</td>\n",
       "      <td>True</td>\n",
       "      <td>NRXN1</td>\n",
       "      <td>risk</td>\n",
       "      <td>Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S36895</td>\n",
       "      <td>SCZ</td>\n",
       "      <td>22q11.21</td>\n",
       "      <td>chr22</td>\n",
       "      <td>18731658</td>\n",
       "      <td>19023031</td>\n",
       "      <td>DEL</td>\n",
       "      <td>291373</td>\n",
       "      <td>LOC102724770, DGCR6, LOC102724788, PRODH</td>\n",
       "      <td>PRODH</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>risk</td>\n",
       "      <td>Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S36933</td>\n",
       "      <td>SCZ</td>\n",
       "      <td>2p16.3</td>\n",
       "      <td>chr2</td>\n",
       "      <td>50270373</td>\n",
       "      <td>50287511</td>\n",
       "      <td>DEL</td>\n",
       "      <td>17138</td>\n",
       "      <td>NRXN1</td>\n",
       "      <td>NRXN1</td>\n",
       "      <td>True</td>\n",
       "      <td>NRXN1</td>\n",
       "      <td>risk</td>\n",
       "      <td>Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S36952</td>\n",
       "      <td>SCZ</td>\n",
       "      <td>2p16.3</td>\n",
       "      <td>chr2</td>\n",
       "      <td>50814739</td>\n",
       "      <td>50830450</td>\n",
       "      <td>DEL</td>\n",
       "      <td>15711</td>\n",
       "      <td>NRXN1</td>\n",
       "      <td>NRXN1</td>\n",
       "      <td>True</td>\n",
       "      <td>NRXN1</td>\n",
       "      <td>risk</td>\n",
       "      <td>Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>S36998</td>\n",
       "      <td>SCZ</td>\n",
       "      <td>1q21.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>145544126</td>\n",
       "      <td>145808812</td>\n",
       "      <td>DEL</td>\n",
       "      <td>264686</td>\n",
       "      <td>GPR89A, PDZK1, CD160, RNF115</td>\n",
       "      <td>CD160</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>risk</td>\n",
       "      <td>Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>S37004</td>\n",
       "      <td>SCZ</td>\n",
       "      <td>1q21.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>145426050</td>\n",
       "      <td>146052017</td>\n",
       "      <td>DUP</td>\n",
       "      <td>625967</td>\n",
       "      <td>GPR89A, PDZK1, CD160, RNF115, POLR3C, NUDT17, ...</td>\n",
       "      <td>CD160, ANKRD34A</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>risk</td>\n",
       "      <td>Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>S37045</td>\n",
       "      <td>BD</td>\n",
       "      <td>22q11.21</td>\n",
       "      <td>chr22</td>\n",
       "      <td>19335594</td>\n",
       "      <td>19348367</td>\n",
       "      <td>DEL</td>\n",
       "      <td>12773</td>\n",
       "      <td>HIRA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>risk</td>\n",
       "      <td>Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>S37122</td>\n",
       "      <td>BD</td>\n",
       "      <td>22q11.21</td>\n",
       "      <td>chr22</td>\n",
       "      <td>18131010</td>\n",
       "      <td>18142544</td>\n",
       "      <td>DUP</td>\n",
       "      <td>11534</td>\n",
       "      <td>TUBA8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protective</td>\n",
       "      <td>Uncertain significance</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID Phenotype  Cytoband Chromosome      Start        End Type  \\\n",
       "0  S36774       SCZ  22q11.21      chr22   18731658   21110696  DEL   \n",
       "1  S36820    UHR-NC    2p16.3       chr2   50672899   50921973  DEL   \n",
       "2  S36895       SCZ  22q11.21      chr22   18731658   19023031  DEL   \n",
       "3  S36933       SCZ    2p16.3       chr2   50270373   50287511  DEL   \n",
       "4  S36952       SCZ    2p16.3       chr2   50814739   50830450  DEL   \n",
       "5  S36998       SCZ    1q21.1       chr1  145544126  145808812  DEL   \n",
       "6  S37004       SCZ    1q21.1       chr1  145426050  146052017  DUP   \n",
       "7  S37045        BD  22q11.21      chr22   19335594   19348367  DEL   \n",
       "8  S37122        BD  22q11.21      chr22   18131010   18142544  DUP   \n",
       "\n",
       "   Overlap with Marshal (bp)  \\\n",
       "0                    1018342   \n",
       "1                     249074   \n",
       "2                     291373   \n",
       "3                      17138   \n",
       "4                      15711   \n",
       "5                     264686   \n",
       "6                     625967   \n",
       "7                      12773   \n",
       "8                      11534   \n",
       "\n",
       "                                               Genes  \\\n",
       "0  LOC102724770, DGCR6, LOC102724788, PRODH, DGCR...   \n",
       "1                                              NRXN1   \n",
       "2           LOC102724770, DGCR6, LOC102724788, PRODH   \n",
       "3                                              NRXN1   \n",
       "4                                              NRXN1   \n",
       "5                       GPR89A, PDZK1, CD160, RNF115   \n",
       "6  GPR89A, PDZK1, CD160, RNF115, POLR3C, NUDT17, ...   \n",
       "7                                               HIRA   \n",
       "8                                              TUBA8   \n",
       "\n",
       "                                         Brain_genes  is_rare Genes (Marshal)  \\\n",
       "0  PRODH, ESS2, SEPTIN5, GP1BB, ARVCF, RTN4R, PI4...     True             NaN   \n",
       "1                                              NRXN1     True           NRXN1   \n",
       "2                                              PRODH    False             NaN   \n",
       "3                                              NRXN1     True           NRXN1   \n",
       "4                                              NRXN1     True           NRXN1   \n",
       "5                                              CD160     True             NaN   \n",
       "6                                    CD160, ANKRD34A     True             NaN   \n",
       "7                                                NaN    False             NaN   \n",
       "8                                                NaN    False             NaN   \n",
       "\n",
       "  Status (Marshal)          Classification  \n",
       "0             risk              Pathogenic  \n",
       "1             risk  Uncertain significance  \n",
       "2             risk  Uncertain significance  \n",
       "3             risk  Uncertain significance  \n",
       "4             risk  Uncertain significance  \n",
       "5             risk  Uncertain significance  \n",
       "6             risk  Uncertain significance  \n",
       "7             risk  Uncertain significance  \n",
       "8       protective  Uncertain significance  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge to find matching rows\n",
    "matching_df = pd.merge(cnv_df, marshal_df, on=[\"Type\", \"Cytoband\"], copy=False)\n",
    "\n",
    "# Calculate overlap\n",
    "overlap_start = matching_df[[\"Start_x\", \"Start_y\"]].max(axis=1)\n",
    "overlap_end = matching_df[[\"End_x\", \"End_y\"]].min(axis=1)\n",
    "matching_df[\"Overlap with Marshal (bp)\"] = (overlap_end - overlap_start).clip(lower=0)\n",
    "\n",
    "# Filter out CNVs with zero overlap\n",
    "matching_df = matching_df[matching_df[\"Overlap with Marshal (bp)\"] > 0]\n",
    "\n",
    "# Select columns and rename them\n",
    "matching_df = matching_df[\n",
    "    [\n",
    "        \"ID\",\n",
    "        \"Phenotype_1\",\n",
    "        \"Cytoband\",\n",
    "        \"Chromosome\",\n",
    "        \"Start_x\",\n",
    "        \"End_x\",\n",
    "        \"Type\",\n",
    "        \"Overlap with Marshal (bp)\",\n",
    "        \"All protein coding genes\",\n",
    "        \"Brain_genes\",\n",
    "        \"is_rare\",\n",
    "        \"Gene\",\n",
    "        \"Status\",\n",
    "        \"Classification\",\n",
    "    ]\n",
    "].rename(\n",
    "    columns={\n",
    "        \"Phenotype_1\": \"Phenotype\",\n",
    "        \"Start_x\": \"Start\",\n",
    "        \"End_x\": \"End\",\n",
    "        \"All protein coding genes\": \"Genes\",\n",
    "        \"Gene\": \"Genes (Marshal)\",\n",
    "        \"Status\": \"Status (Marshal)\",\n",
    "    }\n",
    ")\n",
    "matching_df.reset_index(drop=True, inplace=True)\n",
    "matching_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Additionally, we reviewed CNVs classified as \"Pathogenic\" or \"Likely pathogenic\" with caution, as they may be related to genetic disorders but not necessarily to psychosis. These include: ```Xp22.31``` (1 UHR-NC), ```9p21.3``` (1 UHR-C), ```16p11.2``` (2 SCZ), ```15q11.2``` (1 SCZ), ```8q24.3``` (1 UHR-NC), and ```8q22.1-22.2-22.3``` (1 UHR-NC). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>Cytoband</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Type</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Brain_genes</th>\n",
       "      <th>is_rare</th>\n",
       "      <th>Classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S36747</td>\n",
       "      <td>UHR-NC</td>\n",
       "      <td>Xp22.31</td>\n",
       "      <td>chrX</td>\n",
       "      <td>6535213</td>\n",
       "      <td>8168775</td>\n",
       "      <td>DEL</td>\n",
       "      <td>PUDP, STS, VCX, PNPLA4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S36774</td>\n",
       "      <td>SCZ</td>\n",
       "      <td>22q11.21</td>\n",
       "      <td>chr22</td>\n",
       "      <td>18731658</td>\n",
       "      <td>21110696</td>\n",
       "      <td>DEL</td>\n",
       "      <td>LOC102724770, DGCR6, LOC102724788, PRODH, DGCR...</td>\n",
       "      <td>PRODH, ESS2, SEPTIN5, GP1BB, ARVCF, RTN4R, PI4...</td>\n",
       "      <td>True</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S36795</td>\n",
       "      <td>UHR-C</td>\n",
       "      <td>9p21.3</td>\n",
       "      <td>chr9</td>\n",
       "      <td>21959044</td>\n",
       "      <td>21974294</td>\n",
       "      <td>DEL</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S36908</td>\n",
       "      <td>SCZ</td>\n",
       "      <td>16p11.2</td>\n",
       "      <td>chr16</td>\n",
       "      <td>29599805</td>\n",
       "      <td>30187754</td>\n",
       "      <td>DEL</td>\n",
       "      <td>SPN, QPRT, C16orf54, ZG16, KIF22, MAZ, PRRT2, ...</td>\n",
       "      <td>SPN, PRRT2, PAGR1, SEZ6L2, ASPHD1, DOC2A, TLCD3B</td>\n",
       "      <td>True</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S36961</td>\n",
       "      <td>SCZ</td>\n",
       "      <td>15q11.2</td>\n",
       "      <td>chr15</td>\n",
       "      <td>22650252</td>\n",
       "      <td>23122942</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GOLGA6L7, NIPA1, NIPA2, CYFIP1, TUBGCP5</td>\n",
       "      <td>NIPA1</td>\n",
       "      <td>False</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>S36981</td>\n",
       "      <td>UHR-NC</td>\n",
       "      <td>8q24.3</td>\n",
       "      <td>chr8</td>\n",
       "      <td>142414773</td>\n",
       "      <td>145076114</td>\n",
       "      <td>DUP</td>\n",
       "      <td>ADGRB1, ARC, JRK, PSCA, LY6K, THEM6, SLURP1, L...</td>\n",
       "      <td>ADGRB1, ARC, SLURP2, LYNX1, LY6H, GPIHBP1, NRB...</td>\n",
       "      <td>True</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>S36981</td>\n",
       "      <td>UHR-NC</td>\n",
       "      <td>8q22.1-q22.2-q22.3</td>\n",
       "      <td>chr8</td>\n",
       "      <td>93064965</td>\n",
       "      <td>104546553</td>\n",
       "      <td>DUP</td>\n",
       "      <td>FBXO43, POLR2K, SPAG1, RNF19A, ANKRD46, SNX31,...</td>\n",
       "      <td>NCALD, BAALC, RIMS2, CCNE2, POP1, KCNS2</td>\n",
       "      <td>True</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>S37126</td>\n",
       "      <td>SCZ</td>\n",
       "      <td>16p11.2</td>\n",
       "      <td>chr16</td>\n",
       "      <td>29599805</td>\n",
       "      <td>30187754</td>\n",
       "      <td>DEL</td>\n",
       "      <td>SPN, QPRT, C16orf54, ZG16, KIF22, MAZ, PRRT2, ...</td>\n",
       "      <td>SPN, PRRT2, PAGR1, SEZ6L2, ASPHD1, DOC2A, TLCD3B</td>\n",
       "      <td>True</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID Phenotype            Cytoband Chromosome      Start        End Type  \\\n",
       "0  S36747    UHR-NC             Xp22.31       chrX    6535213    8168775  DEL   \n",
       "1  S36774       SCZ            22q11.21      chr22   18731658   21110696  DEL   \n",
       "2  S36795     UHR-C              9p21.3       chr9   21959044   21974294  DEL   \n",
       "3  S36908       SCZ             16p11.2      chr16   29599805   30187754  DEL   \n",
       "4  S36961       SCZ             15q11.2      chr15   22650252   23122942  DEL   \n",
       "5  S36981    UHR-NC              8q24.3       chr8  142414773  145076114  DUP   \n",
       "6  S36981    UHR-NC  8q22.1-q22.2-q22.3       chr8   93064965  104546553  DUP   \n",
       "7  S37126       SCZ             16p11.2      chr16   29599805   30187754  DEL   \n",
       "\n",
       "                                               Genes  \\\n",
       "0                             PUDP, STS, VCX, PNPLA4   \n",
       "1  LOC102724770, DGCR6, LOC102724788, PRODH, DGCR...   \n",
       "2                                             CDKN2A   \n",
       "3  SPN, QPRT, C16orf54, ZG16, KIF22, MAZ, PRRT2, ...   \n",
       "4            GOLGA6L7, NIPA1, NIPA2, CYFIP1, TUBGCP5   \n",
       "5  ADGRB1, ARC, JRK, PSCA, LY6K, THEM6, SLURP1, L...   \n",
       "6  FBXO43, POLR2K, SPAG1, RNF19A, ANKRD46, SNX31,...   \n",
       "7  SPN, QPRT, C16orf54, ZG16, KIF22, MAZ, PRRT2, ...   \n",
       "\n",
       "                                         Brain_genes  is_rare  \\\n",
       "0                                                NaN     True   \n",
       "1  PRODH, ESS2, SEPTIN5, GP1BB, ARVCF, RTN4R, PI4...     True   \n",
       "2                                                NaN    False   \n",
       "3   SPN, PRRT2, PAGR1, SEZ6L2, ASPHD1, DOC2A, TLCD3B     True   \n",
       "4                                              NIPA1    False   \n",
       "5  ADGRB1, ARC, SLURP2, LYNX1, LY6H, GPIHBP1, NRB...     True   \n",
       "6            NCALD, BAALC, RIMS2, CCNE2, POP1, KCNS2     True   \n",
       "7   SPN, PRRT2, PAGR1, SEZ6L2, ASPHD1, DOC2A, TLCD3B     True   \n",
       "\n",
       "      Classification  \n",
       "0         Pathogenic  \n",
       "1         Pathogenic  \n",
       "2  Likely pathogenic  \n",
       "3         Pathogenic  \n",
       "4         Pathogenic  \n",
       "5  Likely pathogenic  \n",
       "6  Likely pathogenic  \n",
       "7         Pathogenic  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter pathogenic CNVs and rename columns\n",
    "pathogenic_df = cnv_df.loc[\n",
    "    cnv_df[\"Classification\"] != \"Uncertain significance\",\n",
    "    [\n",
    "        \"ID\",\n",
    "        \"Phenotype_1\",\n",
    "        \"Cytoband\",\n",
    "        \"Chromosome\",\n",
    "        \"Start\",\n",
    "        \"End\",\n",
    "        \"Type\",\n",
    "        \"All protein coding genes\",\n",
    "        \"Brain_genes\",\n",
    "        \"is_rare\",\n",
    "        \"Classification\",\n",
    "    ],\n",
    "].rename(columns={\"Phenotype_1\": \"Phenotype\", \"All protein coding genes\": \"Genes\"})\n",
    "pathogenic_df.reset_index(drop=True, inplace=True)\n",
    "pathogenic_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The final dataframe was exported as a CSV file for further analysis and interpretation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Concatenate the matching_df and pathogenic_df\n",
    "df = pd.concat([matching_df, pathogenic_df], ignore_index=True)\n",
    "\n",
    "# Drop duplicate rows based on 'ID' and 'Cytoband', then set as index\n",
    "df.drop_duplicates(subset=[\"ID\", \"Cytoband\"], inplace=True)\n",
    "df.set_index([\"ID\", \"Cytoband\"], inplace=True)\n",
    "\n",
    "# Save the final DataFrame to a CSV file\n",
    "df.to_csv(\"../results/pathogenic_cnv_with_marshal.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "everything",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
